pkg/crane/optimize.go

```diff
@@ -0,0 +1,178 @@
+	srcRef, err := name.ParseReference(src, o.name...)
+	if err != nil {
+		return fmt.Errorf("parsing reference %q: %v", src, err)
+	}
+
+	dstRef, err := name.ParseReference(dst, o.name...)
+	if err != nil {
+		return fmt.Errorf("parsing reference for %q: %v", dst, err)
+	}
+
+	logs.Progress.Printf("Optimizing from %v to %v", srcRef, dstRef)
+	desc, err := remote.Get(srcRef, o.remote...)
+	if err != nil {
+		return fmt.Errorf("fetching %q: %v", src, err)
+	}
+
+	switch desc.MediaType {
+	case types.OCIImageIndex, types.DockerManifestList:
+		// Handle indexes separately.
+		if o.platform != nil {
+			// If platform is explicitly set, don't optimize the whole index, just the appropriate image.
+			if err := optimizeAndPushImage(desc, dstRef, prioritize, o); err != nil {
+				return fmt.Errorf("failed to optimize image: %v", err)
+			}
+		} else {
+			if err := optimizeAndPushIndex(desc, dstRef, prioritize, o); err != nil {
+				return fmt.Errorf("failed to optimize index: %v", err)
+			}
+		}
+
+	case types.DockerManifestSchema1, types.DockerManifestSchema1Signed:
+		return errors.New("docker schema 1 images are not supported")
+
+	default:
+		// Assume anything else is an image, since some registries don't set mediaTypes properly.
+		if err := optimizeAndPushImage(desc, dstRef, prioritize, o); err != nil {
+			return fmt.Errorf("failed to optimize image: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func optimizeAndPushImage(desc *remote.Descriptor, dstRef name.Reference, prioritize []string, o options) error {
+	img, err := desc.Image()
+	if err != nil {
+		return err
+	}
+
+	oimg, err := optimizeImage(img, prioritize)
+	if err != nil {
+		return err
+	}
+
+	return remote.Write(dstRef, oimg, o.remote...)
+}
+
+func optimizeImage(img v1.Image, prioritize []string) (v1.Image, error) {
+	cfg, err := img.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	ocfg := cfg.DeepCopy()
+	ocfg.History = nil
+	ocfg.RootFS.DiffIDs = nil
+
+	oimg, err := mutate.ConfigFile(empty.Image, ocfg)
+	if err != nil {
+		return nil, err
+	}
+
+	layers, err := img.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	olayers := make([]mutate.Addendum, 0, len(layers))
+	for i, layer := range layers {
+		olayer, err := tarball.LayerFromOpener(layer.Uncompressed,
+			tarball.WithEstargzOptions(estargz.WithPrioritizedFiles(prioritize)))
+		if err != nil {
+			return nil, err
+		}
+
+		olayers = append(olayers, mutate.Addendum{
+			Layer:     olayer,
+			History:   cfg.History[i],
+			MediaType: types.DockerLayer,
+		})
+	}
+
+	return mutate.Append(oimg, olayers...)
+}
+
+func optimizeAndPushIndex(desc *remote.Descriptor, dstRef name.Reference, prioritize []string, o options) error {
+	idx, err := desc.ImageIndex()
+	if err != nil {
+		return err
+	}
+
+	oidx, err := optimizeIndex(idx, prioritize)
+	if err != nil {
+		return err
+	}
+
+	return remote.WriteIndex(dstRef, oidx, o.remote...)
+}
+
+func optimizeIndex(idx v1.ImageIndex, prioritize []string) (v1.ImageIndex, error) {
```

Comment:
We really need a:

```go
func Map(f func(v1.Image) (v1.Image, error), i v1.ImageOrIndex) (v1.ImageOrIndex, error)
```

Reply:
I believe I have one downstream somewhere 😇

Reply:
https://github.com/mattmoor/mink/blob/4afdeb3398c4d56401d4638ebd5cb2cb85ad12ad/pkg/bundles/map.go#L118

Reply:
Oh yeah you cced me on this earlier. My only issue is the pushing bit, we are missing an abstraction here for an ImageOrIndex thing (strawman here: https://github.com/google/go-containerregistry/pull/835)